kube-core/src/duration.rs

```diff
@@ -0,0 +1,430 @@
+            is_negative: false
+        }
+    }
+}
+
+impl From<Duration> for time::Duration {
+    fn from(Duration {duration, .. }: Duration) -> Self {
+        duration
+    }
+}
+
+impl Duration {
+    /// Returns `true` if this `Duration` is negative.
+    #[inline]
+    #[must_use]
+    pub fn is_negative(&self) -> bool {
+        self.is_negative
+    }
+}
+
+impl fmt::Debug for Duration {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use std::fmt::Write;
+        if self.is_negative {
+            f.write_char('-')?;
+        }
+        fmt::Debug::fmt(&self.duration, f)
+    }
+}
+
+impl fmt::Display for Duration {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use std::fmt::Write;
+        if self.is_negative {
+            f.write_char('-')?;
+        }
+        fmt::Debug::fmt(&self.duration, f)
+    }
+}
+
+impl FromStr for Duration {
+    type Err = ParseError;
+
+    fn from_str(mut s: &str) -> Result<Self, Self::Err> {
+        // implements the same format as
+        // https://cs.opensource.google/go/go/+/refs/tags/go1.20.4:src/time/format.go;l=1589
+
+        fn duration_from_units(val: f64, unit: &str) -> Result<time::Duration, ParseError> {
+            const MINUTE: time::Duration = time::Duration::from_secs(60);
+            // https://cs.opensource.google/go/go/+/refs/tags/go1.20.4:src/time/format.go;l=1573
+            let base = match unit {
+                "ns" => time::Duration::from_nanos(1),
+                // U+00B5 is the "micro sign" while U+03BC is "Greek letter mu"
+                "us" | "\u{00b5}s" | "\u{03bc}s" => time::Duration::from_micros(1),
+                "ms" => time::Duration::from_millis(1),
+                "s" => time::Duration::from_secs(1),
+                "m" => MINUTE,
+                "h" => MINUTE * 60,
+                _ => return Err(ParseError::InvalidUnit),
+            };
+            Ok(base.mul_f64(val))
+        }
+
+        // Go durations are signed. Rust durations aren't. So we need to ignore
+        // this for now.
+        let is_negative = s.starts_with('-');
+        s = s.trim_start_matches('+').trim_start_matches('-');
+
+        let mut total = time::Duration::from_secs(0);
+        while !s.is_empty() {
+            if let Some(unit_start) = s.find(|c: char| c.is_alphabetic()) {
+                let (val, rest) = s.split_at(unit_start);
+                let val = val.parse::<f64>()?;
+                let unit = if let Some(next_numeric_start) = rest.find(|c: char| !c.is_alphabetic()) {
+                    let (unit, rest) = rest.split_at(next_numeric_start);
+                    s = rest;
+                    unit
+                } else {
+                    s = "";
+                    rest
+                };
+                total += duration_from_units(val, unit)?;
+            } else if s == "0" {
+                return Ok(Duration { duration: time::Duration::from_secs(0), is_negative });
```

Comment:
I think this branch could be changed to a `break`, but it seems nicer to lift the zero check into the `while` condition instead. Perhaps less explicit though.